# Семинар 19. Оценки: Байес против классики
Дата: 2017-02-06    
Подготовил: Анисимов Максим

## ЗАДАЧА 17.3
### Условие
Вася и Петя независимо друг от друга прочитали всю Википедию. Вася всего
нашёл 100 опечаток, Петя — 200 опечаток. При этом 80 опечаток оказались найдены и Петей,
и Васей. С помощью ML и MM:
1. Оцените количество опечаток в Википедии
2. Оцените внимательность Васи, то есть вероятность, с которой Вася находит опечатки

Пусть:
+ $n$ - количество отпечаток в Википедии    
+ $X^B$ - количество опечаток, найденных Васей
+ $p$ - вероятность Васи найти опечатку
+ $X^П$ - количество опечаток, найденных Петей
+ $q$ - вероятность Пети исправить опечаток
+ $X^{ВП}$ - количество опечатое, найденных обоими героями

$\hat{n}=?$    
$\hat{p}=?$

### Решение
Предположим, что распределение исправления ошибок является биномиальным.
Тогда $$P(X^B=100 \cap X^П=200 \cap X^{ВП}=80) = C_{n}^{100}*p^{100}*(1-p)^{n-100}*C_{100}^{80}*q^{80}*(1-q)^{20}*C_{120}^{n-100}*q^{120}*(1-q)^{n-220}$$
Как найти $n$, $p$, $q?$

Испробуем два классических метода:
1. Метод максимального правдоподобия (ММП; Method of Likelihood - ML)
2. Метод моментов (ММ; Method of Moments - MM)

#### Метод максимального правдоподобия
Найдем оценки $\hat{q}, \hat{p}, \hat{n}$.
Вся задача сводится к реализации идеи, что наша выборка является максимально правдопдобной, что требудет максимизации функции правдоподобности $L$. В нашем случае:    
$$ L(n,p,q)=P(X^B=100 \cap X^П=200 \cap X^{ВП}=80)=C_{n}^{100}*p^{100}*(1-p)^{n-100}*C_{100}^{80}*q^{80}*(1-q)^{20}*C_{120}^{n-100}*q^{120}*(1-q)^{n-220} \longrightarrow max$$    
Как всегда в таких случаях на помощь приходит логарифм! Найдем максимум $lnL$, ведь от этого значения оценок в максимуме не изменятся.    
$$lnL =  lnC_{n}^{100}+lnp^{100}+ln(1-p)^{n-100}+lnC_{100}^{80}+lnq^{80}+ln(1-q)^{20}+lnC_{120}^{n-100}+lnq^{120}+ln(1-q)^{n-220} \longrightarrow max$$    
$$\dfrac{\partial\ lnL}{\partial\ p}=\frac{100}{\hat{p}}+\frac{100-\hat{n}}{1-\hat{p}}=0 => \hat{p}=\frac{100}{\hat{n}}$$      

$$\dfrac{\partial\ lnL}{\partial\ q}=\frac{80}{\hat{q}}-\frac{20}{1-\hat{q}}\frac{120}{\hat{q}}-\frac{\hat{n}-220}{1-\hat{q}}=0 => \hat{q}=\frac{200}{\hat{n}}$$      
Производную по $n$ брать сложновато, поэтому воспользуемся свойством дискретной функции в ее максимуме, ведь $n$ - количество ошибок, а значит, целое число!
$$\frac{L(n+1,p,q)}{L(n,p,q)}=\frac{(n+1)(1-p)(1-q)}{n-219}=1$$
Отсюда $$\hat{n}=\frac{220}{\hat{q}+\hat{p}+\hat{p}\hat{q}}-1$$    
Благодаря умению дифферецировать, мы уже нашли оптимальные $\hat{p}$ и $\hat{q}$! Осталось только подставить их в выражение для $\hat{n}$ и найти оптимальный $\hat{n}$.    
$$\hat{n}=\frac{200}{\frac{200}{\hat{n}}+\frac{100}{\hat{n}}-\frac{20000}{\hat{n}^{2}}}-1$$     
$$\frac{80\hat{n}^{2}-1970\hat{n}-20000}{300\hat{n}-20000}=0$$      
Далее решим квадратное уравнение в числителе и найдем $\hat{n}$.

In [13]:
a = float(80)
b = float(-19700)
c = float(-20000)
 
discr = b**2 - 4 * a * c;
if discr > 0:
	import math
	x1 = (-b + math.sqrt(discr)) / (2 * a)
	x2 = (-b - math.sqrt(discr)) / (2 * a)
	print("x1 = %.2f \nx2 = %.2f" % (x1, x2))
elif discr == 0:
	x = -b / (2 * a)
	print("x = %.2f" % x)
else:
	print("Корней нет")

n = float()
if x1 >= 0 and x2 <= 0:
    n = x1
    print(n)
elif x2 >= 0 and x1 <= 0:
    n = x2
    print(n)
elif x1 >= 0 and x2 >= 0:
    n1, n2 = x1, x2
    print(n1,n2)
else:
    print("Оптимального $n$ не существует")
print(100/n)
print(200/n)
    

x1 = 247.26 
x2 = -1.01
247.26107704853575
0.4044308194142932
0.8088616388285864


#### Ответ
$\hat{n}=247.26107704853575$    
$\hat{p}=0.4044308194142932$    
$\hat{q}=0.8088616388285864$    

#### Метод моментов
Данный способ решения подразумевает, что мы воспользуемся моментами, чтобы найти значения оценок.
Пусть $z_{1},z_{2},...,z_{n}$ - независимые одинаково распределенные случайные величины.
Тогда $E(Z_{i})|_{\theta=\hat{\theta}}=\bar{Z}$    
Далее запишем таблицы распределения для случайных величин $Z_{i}$, $W_{i}$, $Y_{i}$, где
$Z_{i}$-случайная величина ("индикатор"), принимающая значения 0, если Вася не нашёл i-ую ошибку в Википедии, и 1, если нашел, $W_{i}$ - индикатор Пети, $Y_{i}$ - индикатор Васи и Пети вместе.

| $Z_{i}$  | $0$      |  $1$   |    
|:-----------|:--------:|-------:|
| $P$        | 1-p      |  p     |

| $W_{i}$  | $0$      |  $1$   |
|:-----------|:--------:|-------:|
| $P$        | 1-q      |  q     |

| $Y_{i}$  | $0$      |  $1$   |
|:-----------|:--------:|-------:|
| $P$        | 1-pq      |  pq     |


Наше предположении о биноминальном распределении данных случайных величин даёт свои плоды! Мы можем выписать первые центральные моменты, то есть математические ожидания, для каждой из них, которые равны вероятности успеха найти ошибку:
$$E(Z_i)=p$$
$$E(W_i)=q$$
$$E(Y_i)=pq$$

Теперь найдем оценки математического ожидания $\bar{Z}$, $\bar{W}$, $\bar{Y}$, которые равны доле успехов в n испытаниях: 
 $$\bar{Z}=\frac{100}{n}$$
 $$\bar{W}=\frac{200}{n}$$
 $$\bar{Y}=\frac{80}{n}$$
 
Так как $E(Z_{i})|_{\theta=\hat{\theta}}=\bar{Z}$, остается найти значения $\hat{p}$, $\hat{q}$, $\hat{n}$ из системы уравнений:
\begin{cases}
\hat{p}=\frac{100}{\hat{n}}\\
\hat{q}=\frac{200}{\hat{n}}\\
\hat{p}\hat{q}=\frac{80}{\hat{n}}\\
\end{cases} 
Ответ готов!

#### Ответ
$\hat{p}=0.4$   
$\hat{q}=0.8$    
$\hat{n}=250$

И ничего удивительного, что ответы, полученные с помощью обоих методов, очень близки друг к другу!

## ЗАДАЧА ПРО МАШУ И МЕДВЕДЕЙ
### Байес вступает в бой! Теперь решим другу задачу с помощью байесовского подхода.    
### Условие
Маша прячется от трех медведей в точке $m$. Через какое-то время они начинают вынюхивать Машу, на основе чего предполагают, где она находится.

В данной задаче байесовский подход поможет найти нам функцию распределения $m$.    
Пусть $y_1$,$y_2$,$y_3$ - независимые одинаково распределенные случайные величины, которые означают место, где сильнее всего пахнет Машей для медведей $1$,$2$ и $3$ соответственно.   
Также предположим, что: $$y_i|m \sim N(m;2^2), i=1,2,3$$
$$m \sim N(7;10^2)$$    
Вдобавок, мы обладаем информацией, что $y_1=27$, $y_2=28$, $y_3=23$.

#### Решение
$f(m|y_1,y_2,y_3)=\frac{f(m,y1,y2,y3)}{f(y1,y2,y3)} \propto f(m,y_1,y_2,y_3)=f(y_1,y_2,y_3|m)f(m)=f(y_1|m)f(y_2|m)f(y_3|m)f(m)=\frac{1}{10\sqrt{2\pi}}e^{-\frac{1}{2}(\frac{m-7}{10})^2}*\Pi_{i=1}^{3}[\frac{1}{2\sqrt{2\pi}}e^{-\frac{1}{2}(\frac{y_i-m}{2})^2}] \propto e^{-\frac{1}{2}[(\frac{m-7}{10})^2+(\frac{27-m}{2})^2+(\frac{28-m}{2})^2+(\frac{23-m}{2})^2]} \propto e^{-\frac{1}{2}(\frac{m^2-\frac{3194m}{76}}{\frac{100}{76}})} \propto e^{-\frac{1}{2}[\frac{m-\frac{1957}{76}}{10\sqrt{76}}]^2}$

Исходя из функции нормального распределния, легко можем найти $E(m)$ и $\sigma_m$.


#### Ответ
$f(m|y_1,y_2,y_3) \propto \frac{1}{10\sqrt{2\pi}}e^{-\frac{1}{2}[\frac{m-\frac{1957}{76}}{10\sqrt{76}}]^2}$    
$m(y_1,y_2,y_3) \sim N(25.75; 1.32)$    


### Эту же задачу можно было решить классическими методами!
#### Метод моментов
Например, чтобы найти оценку математического ожидания, надо взять среднее по выборке, то есть:
$$\hat{m}=\frac{\Sigma_{i=1}^{3}y_i}{3}=\frac{27+28+23}{3}=26$$

#### Метод максимального правдоподобия
В этой задаче функция правдоподобия $L(m,y_1,y_2,y_3)=f(y_1,y_2,y_3|m)$. Её мы будем максимизировать для нахождения, например, оценки метматического ожидания.
$$L(m)=e^{-\frac{1}{2}[(\frac{27-m}{2})^2+(\frac{28-m}{2})^2+(\frac{23-m}{2})^2} \longrightarrow max$$
$$lnL(m)=-\frac{1}{2}[(\frac{27-m}{2})^2+(\frac{28-m}{2})^2+(\frac{23-m}{2})^2] \longrightarrow max$$
Отсюда получаем, что $\hat{m}=26$.


In [14]:
import pylab
from scipy.stats import norm
m = np.linspace(start=10,stop=40,num=10000)
y = norm.pdf(m, loc=25.75, scale=1.32)  
pylab.plot(m,y)
pylab.show()